### PR TITLE
Enabling upgrade testing during acceptance pipeline

### PR DESCRIPTION
**Pull request by zmc**
This is related to [rhbz#1999689](https://bugzilla.redhat.com/show_bug.cgi?id=1999689), but also requires an MR to our Jenkins pipeline to resolve the BZ.

### CODE REVIEW

**Comment by zmc** [comment]
@petr-balogh @clacroix12 please take a look when you can

File: ocs_ci/framework/tests/test_main.py
Comment (clacroix12) [inline ocs_ci/framework/tests/test_main.py:65]: Why does the main function return a ValueError with these args now? Shouldn't this be a legitimate use case?

**Comment by clacroix12** [comment]
Adding `Needs Deployment Testing` as this changes functionality in the main function.

File: .functional_ci_setup.py
Comment (zmc) [reply .functional_ci_setup.py:71]: I could go either way on this. If you prefer, I can use the constant. I'd just have to re-test - but on the other hand, it works as-is for our use case.

**Comment by zmc** [comment]
@clacroix12 thanks for the review and the trigger. The job failed to deploy, but appears to have selected the version properly.

File: .functional_ci_setup.py
Comment (clacroix12) [reply .functional_ci_setup.py:71]: It's not required,  I just wanted to mention it. If you're fine with the current logic then you can keep it as is.

**Comment by clacroix12** [comment]
> @clacroix12 thanks for the review and the trigger. The job failed to deploy, but appears to have selected the version properly.

Np! As getting the correct version and loading the config are what changed in the main function I think that this is verified. The deployment failed due to a health warning but that doesn't seem to be related to these changes at all. Marking as verified.

**Comment by petr-balogh** [comment]
Hey @zmc , can you please join tomorrows automation meeting to quickly syncup on this one?

I can approve it now but I didn't get it fully.
In the our scenario we are specifying --ocs-version and --upgrade-ocs-version where --upgrade-ocs-version specifying config value:
`upgrade_ocs_version` under UPGRADE config section.
This will allow ocs-ci know on which version it's starting and to which version you upgrading.

I don't see in this PR or in MR any of this upgrade-ocs-version set.
I guess you should upgrade from previous version let's say for building ocs 4.10 build, you should trigger upgrade job for upgrade from 4.9 to 4.10 right?

Also we are usually deploying GAed version of previous version when previous version is GAed.
So we are building cluster with `--live-deploy` and then we are upgrading to `--upgrade-ocs-registry-image` I also don't see where exactly you are defining --live-deploy or `live_deployment` config value.

Also you need to count on the usecase when previous version is not GAed linke before 4.10 is GAed you will probably start already producing 4.11 builds.

**Comment by zmc** [comment]
> Hey @zmc , can you please join tomorrows automation meeting to quickly syncup on this one?

I should be able to join for a few minutes, yeah

> I can approve it now but I didn't get it fully. In the our scenario we are specifying --ocs-version and --upgrade-ocs-version where --upgrade-ocs-version specifying config value: `upgrade_ocs_version` under UPGRADE config section. This will allow ocs-ci know on which version it's starting and to which version you upgrading.

I set [`upgrade_ocs_registry_image` here](https://github.com/red-hat-storage/ocs-ci/pull/5298/files#diff-fa3f27d045dc38c6cad83a31ae9e31a87d1e27ddf89df0567909063584eb441cR126-R129), which means [we don't need `upgrade_ocs_version`](https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/ocs/ocs_upgrade.py#L248-L252).

> I don't see in this PR or in MR any of this upgrade-ocs-version set. 

See above 😊

> I guess you should upgrade from previous version let's say for building ocs 4.10 build, you should trigger upgrade job for upgrade from 4.9 to 4.10 right?

Yes, that's what this PR does; see [here](https://github.com/red-hat-storage/ocs-ci/pull/5298/files#diff-fa3f27d045dc38c6cad83a31ae9e31a87d1e27ddf89df0567909063584eb441cR114-R120)

> Also we are usually deploying GAed version of previous version when previous version is GAed. So we are building cluster with `--live-deploy` and then we are upgrading to `--upgrade-ocs-registry-image` I also don't see where exactly you are defining --live-deploy or `live_deployment` config value.
> 
> Also you need to count on the usecase when previous version is not GAed linke before 4.10 is GAed you will probably start already producing 4.11 builds.

This is where it gets a little more tricky. If we want to test version N, and N-1 is GA, then we can use `live_deployment`, but if it is not GA then we can't. On the other hand, we can pull GA bits from the same place we get pre-release bits - that's why this behaves the way it does currently.

It would be ideal if the framework itself had a way of determining whether N-1 was live or not, and use `live_deployment` where possible.

**Comment by petr-balogh** [comment]
@zmc 

OCS-CI framework allows you to deploy basically whatever you explicitly define. From internal build, stage or live content.
Based on configuration user provides to the framework.

There are many possibilities for various upgrade scenarios QE can tests. See below:

- Internal build to internal build (y-stream and z-stream also both automatic and manual approval strategy)
- GAed version to Internal build (z-stram, y-stream upgrades)
- GAed version to GAed version (y-stream, z-stream is possible with manual approval strategy - and cluster needs to be created before release - once latest z-stream is released you cannot install x.y.z-1 from live content) 
- Stage to Stage - (y-stream, z-stream only when cluster is prepared before push to the stage and with manual approval strategy - similar like live upgrade testing)

In order to find GA version it requires to have cluster with intact catalog source - the original redhat-operators.
As catalog source is now named redhat-operators for internal builds as well as origin one it's hard to figure out the GAed version if user touched catalog source to point for example to internal build.
Currently it's possible to find such version from packageManifest but it depends  (as I mentiond above) on catalogSource and require to have OCP cluster installed with intact catalogSource redhat-operators.

Maybe you as build team knows if we have some service how to find the other way the released OCS/ODF versions?

Because we have such many various combinations, the upgrade process is now only for explicit logic where user needs to specify what they would like to deploy and then for upgrade it's also based on configurations.

I think that at the time you are triggering the build you better know if the previous version is GAed or not to select this by some checkbox to deploy the live version of X.Y-1 will be quite simple and by default if there is such possibility you should test those upgrade from GAed version to internal build from X.Y-1 to X.Y .

I will approve this PR for now, but still thing that by default you should just install X.Y-1 From GAed version whenever possible and only in the early builds of future version when the N-1 is not GAed yet to install from internal build.

**Review by petr-balogh — APPROVED** [review]

**Comment by zmc** [comment]
Thanks for the review. I agree that it's probably best to to source the initial deployment from a live build. I'm glad we're not blocking on that, though, as this work was never meant to obsolete QE executing and analyzing upgrade tests on its own.

Because this job is executed via automation on build completion, we don't have a way to interactively make different choices around things like this. We'll need that decision to be automated as well. I don't know the best way to answer the question of "what is the latest release in stream X.Y" - I'll look into it.

**Pull request merged**
